src/Language/Spectacle/RTS/Registers.hs

```diff
@@ -0,0 +1,91 @@
+-- | Runtime state used to implement call-by-need evaluation of closures and variable substitution.
+--
+-- @since 0.1.0.0
+module Language.Spectacle.RTS.Registers
+  ( RuntimeState (RuntimeState, plains, primes, callStack),
+    Registers (Registers, unRegisters),
+    emptyRegisters,
+    getRegister,
+    setRegister,
+    setThunk,
+    Thunk (Thunk, Evaluated, Unchanged),
+  )
+where
+
+import Data.Type.Rec (HasSel (getRecT, setRecT), Name, Rec, RecT, fieldMap, type (#), type (.|))
+import Language.Spectacle.Exception.RuntimeException (RuntimeException)
+import Language.Spectacle.Lang (Lang)
+import Language.Spectacle.Syntax.Error.Internal (Error)
+import Language.Spectacle.Syntax.NonDet.Internal (NonDet)
+import Language.Spectacle.Syntax.Plain.Internal (Plain)
+import Language.Spectacle.Syntax.Prime.Internal (Prime)
+
+-- -------------------------------------------------------------------------------------------------
+
+-- | Internal state used by variable substitution and variable relations.
+--
+-- * 'plains' is a record relating plain variables in @ctx@ to values in the previous frame of time.
+-- * 'primes' is a record relating primed or "time" variables in @ctx@ to their values in the next
+-- frame of time.
+-- * 'callStack' tracks the substitution of primed variables which is used guard against a cyclic
+-- variable relations.
+--
+-- @since 0.1.0.0
+data RuntimeState ctx = RuntimeState
+  { plains :: Rec ctx
+  , primes :: Registers ctx
+  , callStack :: [String]
+  }
+
+-- | A record of 'Thunk's.
+--
+-- @since 0.1.0.0
+newtype Registers ctx = Registers
+  {unRegisters :: RecT (Thunk ctx) ctx}
+
+-- | Construct a 'Registers' of unrelated primed variables.
+--
+-- @since 0.1.0.0
+emptyRegisters :: Rec ctx -> Registers ctx
+emptyRegisters = Registers . fieldMap (const Unchanged)
+
+-- | Retrieves the value of the variable named @s@ in 'Registers' as a 'Thunk'.
+--
+-- @since 0.1.0.0
+getRegister :: s # a .| ctx => Name s -> Registers ctx -> Thunk ctx a
```

Comment:
Just a thought: unless there's precedent for using `#` in the literature, or you want to use as few characters as possible, you could rename `#` to `:::` to resemble more familiar syntax. Or `.::`, to be consistent with the `.|` operator.